### ISSUE TITLE
RegEx flags in patterns are not honored when checking examples

### ISSUE THREAD

**Issue by wedi** [stoplightio/spectral#2030]
**Describe the bug**
The following schema contains a valid ECMA 262 regular expression with a [matching example](https://regex101.com/r/5Mm1uZ/1) which gets rejected by spectral.

```yaml
reference:
  type: string
  pattern: '/\P{Cc}+ [1-9]?[0-9]+/u'
  example: 爱 1
```
```sh
53:14  error  oas3-valid-schema-example  "example" property must match pattern "/\P{Cc}+ [1-9]?[0-9]+/u"  properties.reference.example
```

You fixed this issue for [patterns in rule sets](https://github.com/stoplightio/spectral/issues/242).
Some time ago you rejected to [allow slashes for patterns](https://github.com/stoplightio/spectral/issues/1115).
Spectral does not complain about the flags (or slashes) themselves but does not honor them when checking examples.

**Expected behavior**
Honor [flags](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp) (which you mention as a [comment in spectral](https://github.com/stoplightio/spectral/blob/2f26ed672c98cbd308fc1dd1daff803d36dcd42d/packages/functions/src/pattern.ts#L22)) when checking examples.

My actual use case are the [unicode character classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Unicode_Property_Escapes) that require the `u` flag.

**Environment:**
```
$ node ./node_modules/@stoplight/spectral-cli/dist/index.js --version
6.1.0
$ node --version
v17.4.0
```

**Comment by P0lip** [comment]
`pattern` keyword doesn't allow flags to be specified, so the aforementioned behavior is somewhat expected.

Overall regular expressions in [JSON Schema](https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.6.4) _should_  (should means it's not a hard requirement) built with the Unicode flag, but we disabled it over here https://github.com/stoplightio/spectral/pull/1787

The function you brought up is not used for examples. It's just a plain function you can use in your own rules.

**Comment by wedi** [comment]
Reading up the standard more closely, I see now that the term `pattern` used in JSON schema is a defined term that describes just a part of the ECMA 262 `RegExp`. So the regex I used is indeed a ECMA 262 `RegExp` but you can only use the `pattern` part in `/<pattern>/<flags>` or `RegExp(pattern, flags)`.

Although it's a pitty I understand your reasoning in #1781.

I forked the project to make a pull request to the documentation but I'm struggling to find a suitable spot for this information. The example rule is responsible for a lot of formats so I'm hesitant to add it there but I have no better idea.

**Comment by P0lip** [comment]
> Reading up the standard more closely, I see now that the term pattern used in JSON schema is a defined term that describes just a part of the ECMA 262 RegExp. So the regex I used is indeed a ECMA 262 RegExp but you can only use the pattern part in /<pattern>/<flags> or RegExp(pattern, flags).

That's correct.

> Although it's a pitty I understand your reasoning in #1781.

Nothing is set in stone forever. I'm happy to revisit the standpoint.
Without doubt, having `unicode` enabled would be much better and I'd love to have that turned on, but the problem is that some people would need to be more careful about the regular expressions they craft.
At least we could make that option configurable, so folks that do intend to have the unicode support can do so.


> I forked the project to make a pull request to the documentation but I'm struggling to find a suitable spot for this information. The example rule is responsible for a lot of formats so I'm hesitant to add it there but I have no better idea.

Could maybe add a new file in `docs/reference`?
